File: level2/ex62.js
function solution(expression) {
    const isDigit = src => !isNaN(src);
    
    function split(src) {
        const dest = [];
        let num = '';
        
        for (const value of src) {
            if (isDigit(value)) {
                num += value;
            } else {
                dest.push(parseInt(num));
                dest.push(value);
                num = '';
            }
        }
        
        return [...dest, parseInt(num)];
    };
    
    function calculate(src, op) {
        const dest = [];
        
        for (let i = 0; i < src.length; i++) {
            if (src[i] === op) {
                const num1 = dest.pop();
                const num2 = src[i+1];
                dest.push(eval(num1+op+num2));
            } else if (src[i-1] === op) {
                continue;
            } else {
                dest.push(src[i]);
            }
        }
        
        return dest;
    };
    
    function findMax(src) {
        const result = [];
        const perms = [
            ['+', '-', '*'],
            ['+', '*', '-'],
            ['-', '+', '*'],
            ['-', '*', '+'],
            ['*', '+', '-'],
            ['*', '-', '+'],
        ];
        
        for (const perm of perms) {
            let dest = split(src);
            dest = calculate(dest, perm[0]);
            dest = calculate(dest, perm[1]);
            dest = calculate(dest, perm[2]);
            result.push(Math.abs(dest[0]));
        }
        
        return Math.max(...result);
    }
    
    return findMax(expression);
}

File: level2/ex62.py
from itertools import permutations

def solution(expression):
    def split(src):
        dest, num = [], ''
        
        for value in src:
            if value.isdigit():
                num += value
            else:
                dest.append(int(num))
                dest.append(value)
                num = ''
        
        return dest + [int(num)]
    
    def calculate(src, op):
        dest = []
        index = [i for i, value in enumerate(src) if value == op]

        for i, value in enumerate(src):
            if i in index:
                if op == '+':
                    dest[-1] += src[i+1]
                if op == '-':
                    dest[-1] -= src[i+1]
                if op == '*':
                    dest[-1] *= src[i+1]
            elif i-1 in index:
                continue
            else:
                dest.append(value)
            
        return dest
    
    def find_max(src, ops=['+', '-', '*']):
        result = []
        perms = permutations(ops)
        
        for perm in perms:
            dest = split(src)
            dest = calculate(dest, perm[0])
            dest = calculate(dest, perm[1])
            dest = calculate(dest, perm[2])
            result.append(abs(dest[0]))

        return max(result)
    
    return find_max(expression)